src/ast/passes/named_param.cpp

```diff
@@ -0,0 +1,128 @@
+#include <algorithm>
+
+#include "ast/ast.h"
+#include "ast/context.h"
+#include "ast/passes/named_param.h"
+#include "ast/visitor.h"
+#include "clang_parser.h"
+#include "driver.h"
+#include "util/strings.h"
+
+namespace bpftrace::ast {
+
+class NamedParamPass : public Visitor<NamedParamPass> {
+public:
+  NamedParamPass(ASTContext &ast, BPFtrace &bpftrace)
+      : ast_(ast), bpftrace_(bpftrace) {};
+
+  using Visitor<NamedParamPass>::visit;
+  void visit(Expression &expr);
+
+  std::unordered_map<std::string, globalvars::GlobalVarValue> used_args;
+  NamedParamDefaults defaults;
+
+private:
+  ASTContext &ast_;
+  BPFtrace &bpftrace_;
+};
+
+void NamedParamPass::visit(Expression &expr)
+{
+  auto *call = expr.as<Call>();
+  if (!call || call->func != "getopt") {
+    Visitor<NamedParamPass>::visit(expr);
+    return;
+  }
+
+  auto *arg_name = call->vargs.at(0).as<String>();
+  if (!arg_name) {
+    call->vargs.at(0).node().addError()
+        << "First argument to 'getopt' must be a string literal.";
+    return;
+  }
+
+  bool is_arg_error = false;
+  if (call->vargs.size() == 2) {
+    if (!call->vargs.at(1).as<Integer>() &&
+        !call->vargs.at(1).as<NegativeInteger>() &&
+        !call->vargs.at(1).as<String>() && !call->vargs.at(1).as<Boolean>()) {
+      is_arg_error = true;
+    }
+    if (is_arg_error) {
+      call->vargs.at(1).node().addError()
```

Comment:
Nit: you could roll this into the `else` case below and it will simplify the control flow a bit. Up to you.